flocker/node/agents/blockdevice.py

```diff
@@ -1468,6 +1540,94 @@ def _get_system_mounts(self, volumes, compute_instance_id):
             if FilePath(partition.device) in device_to_dataset_id
         }
 
+    def _upgrade_1_0_0_filesystem(self, api, volumes):
+        """
+        Upgrade any old, Flocker 1.0.0-created filesystems.
+
+        Examine each volume attached to this host.  Determine if the filesystem
+        on the volume has the corresponding dataset_id recorded in its UUID.
+        This is how filesystems are created post-1.0.0 (see FLOC-2369).  If the
+        filesystem does not have this information recorded and the filesystem
+        is not mounted, put it there.  If the filesystem is mounted we cannot
+        modify it.  We will fix it when we notice it is unmounted.
+
+        The intent is for this to cause Flocker 1.0.0-created volumes to look
+        just like they would have if they had been created by Flocker >1.0.0.
+
+        :param IBlockDeviceAPI api: The storage driver responsible for the
+            volumes.
+        :param list volumes: The ``BlockDeviceVolumes`` that exist and will be
+            considered for upgrade.
+        """
+        mounted = {
+            FilePath(partition.device)
+            for partition
+            in psutil.disk_partitions()
+        }
+
+        compute_instance_id = api.compute_instance_id()
+        for volume in _attached_volumes(compute_instance_id, volumes):
+            try:
+                get_device_for_dataset_id(volume.dataset_id)
+                # If we found the dataset_id like this the filesystem is
+                # sufficiently new to not need any changes.  Proceed to the
+                # next volume.
+                continue
+            except KeyError:
+                # There is no filesystem on the volume because volume setup
```

Comment:
> Big picture point: upgrades shouldn't cause slow downs ever when not upgrading

Not sure I agree with this.  We need to be pragmatic.  Some slowdowns don't matter.  This **particular** slowdown may be too egregious, though.

The other idea I considered (instead of asking get_device_path first) was only trying once per process-lifetime per dataset_id.  This would still be slow but only infrequently.

Asking get_device_path first might be the right thing to do.  I'm not sure.  I'm concerned it will produce the wrong results (detach/attach seems more likely to be reliable).  But we already depend on it being right after the detach/attach so that may not be a valid concern.  The real impediment to that approach is how to write tests for the extra cases.